core/src/main/scala/org/apache/spark/storage/ShuffleBlockFetcherIterator.scala

```diff
@@ -261,14 +281,58 @@ final class ShuffleBlockFetcherIterator(
             results.put(new SuccessFetchResult(BlockId(blockId), infoMap(blockId)._2,
               address, infoMap(blockId)._1, buf, remainingBlocks.isEmpty))
             logDebug("remainingBlocks: " + remainingBlocks)
+            enqueueDeferredFetchRequestIfNecessary()
           }
         }
         logTrace(s"Got remote block $blockId after ${Utils.getUsedTimeNs(startTimeNs)}")
       }
 
       override def onBlockFetchFailure(blockId: String, e: Throwable): Unit = {
         logError(s"Failed to get block(s) from ${req.address.host}:${req.address.port}", e)
-        results.put(new FailureFetchResult(BlockId(blockId), infoMap(blockId)._2, address, e))
+        val (size, mapIndex) = infoMap(blockId)
+        e match {
+          // SPARK-27991: Catch the Netty OOM and set the flag `isNettyOOMOnShuffle` (shared among
+          // tasks) to true as early as possible. Unless there's no in-flight requests, the pending
+          // fetch requests won't be raised afterwards until the flag is set to false on:
+          // 1) the Netty free memory >= average remote block size - we'll check this whenever
+          //    there's a fetch request succeeds.
+          // 2) the number of in-flight requests becomes 0 - we'll check this in `fetchUpToMaxBytes`
+          //    whenever it's invoked.
+          // Although Netty memory is shared across multiple modules, e.g., shuffle, rpc, the flag
+          // only takes effect for the shuffle due to the implementation simplicity concern.
+          // And we'll buffer the consecutive block failures caused by the OOM error until there's
+          // no remaining blocks in the current request. Then, we'll package these blocks into
```

Comment:
really? we package these request when a next `onBlockFetchSuccess` happens, which is not end of the request, right?